src/bloomfilter.cpp

```diff
@@ -0,0 +1,206 @@
+#include "include/bloomfilter.h"
+#include "DuckLogger.h"
+
+
+BloomFilter::BloomFilter(int numSectors, int numHashes, int bitsPerSector, int maxMsgs) {
+    logdbg(numSectors);
+    this->numSectors = numSectors; 
+    logdbg("added to BF");
+    this->numHashes = numHashes;
+    logdbg("added to BF");
+    this->bitsPerSector = bitsPerSector;
+    logdbg("added to BF");
+    this->activeFilter = 1;
+    this->maxMsgs = maxMsgs;
+    this->nMsg = 0;
+
+    logdbg("initialize bloom filter 1");
+    // Initialize the bloom filters, fill with 0's
+    this->filter1 = new unsigned int[this->numSectors];
+    logdbg_f("Filter 1 address %p\n", this->filter1);
+    if (this->filter1 == NULL) {
+        logdbg("Memory allocation for Bloom Filter 1 failed!\n");
+        exit(0);
+    }
+    for (int n = 0; n < this->numSectors; n++) {
+        this->filter1[n] = 0;
+    }
+    logdbg_f("Initialized BF1, %d slots, %d this->numSectors\n", numSectors, this->numSectors);
+    
+    logdbg("initialize bloom filter 2");
+    this->filter2 = new unsigned int[this->numSectors];
+    logdbg_f("Filter 2 address %p\n", this->filter2);
+    if (this->filter2 == NULL) {
+        logdbg("Memory allocation for Bloom Filter 2 failed!\n");
+        exit(0);
+    }
+    for (int n = 0; n < this->numSectors; n++) {
+        this->filter2[n] = 0;
+    }
+    logdbg_f("Initialized BF2, %d slots, %d this->numSectors\n", numSectors, this->numSectors);
+
+    logdbg("initialize random seeds");
+    //get random seeds for hash functions 
+    int* Seeds = new int[numHashes];
+    if (Seeds == NULL) {
+        logdbg("Memory allocation for seeds failed!\n");
+        exit(0);
+    }
+    srand(time(NULL));
+    int i;
+    for (i = 0; i < numHashes; i++){
+        int r = rand();
+        // Ensure we have no repeat seeds
+        int seedCollision = 1;
+        while (seedCollision == 1) {
+            seedCollision = 0;
+            for (int i = 0; i < numHashes; i++){
+                if (Seeds[i] == r) {
+                    seedCollision = 1;
+                }
+            }
+            if (seedCollision == 1){
+                r = rand();
+            }
+        }
+        Seeds[i] = r;
+    }
+    for (i = 0; i < numHashes; i++){
+        logdbg_f("random seed %d: %d\n", i+1, Seeds[i]);
+    }
+    this->Seeds = Seeds;
+
+    logdbg("bloom_init end");
+}
+
+BloomFilter::~BloomFilter()
+{
+    delete this->Seeds;
+    delete this->filter2;
+    delete this->filter1;
+}
+
+void BloomFilter::set_hash_results(unsigned char* msg, int msgSize,
+    std::unique_ptr<unsigned int[]> & hashResults
+) {
+    for (int i = 0; i < this->numHashes; i++){
+        unsigned int hashResult = djb2Hash(msg, this->Seeds[i], msgSize);
+        unsigned int totalBitSize = this->numSectors * this->bitsPerSector;
+        hashResult = hashResult % totalBitSize;

```

Comment:
This is the place we fixed the bloom filter math. The previous version was something like:

```
hashResult = hashResult % this->numSectors;
```

Terminology:

We're using "sector" to mean one `unsigned int`.

The previous code for this part was basically treating each sector (unsigned int) as a bit map of size one. Other parts of the code treat each sector as a bit map of multiple bits. (The number of bits per sector is the third parameter of the BloomFilter constructor.) Using all the bits in a sector means basically doing `sizeof(unsigned int)` to see how many bytes are in an unsigned int. Then multiply by the number of bits per byte, and you've got the maximum number of bits in a sector. That's 32-bits on my Heltec V2.

Reply:
Testing this branch now

Reply:
Would a hardware running on an 8-bit system work differently or will it adust?

Reply:
Currently, it is hard-coded. We could use `sizeof()` to get the size based on the device's architecture.